cpp/src/io/parquet/page_delta_decode.cu

```diff
@@ -0,0 +1,164 @@
+  // that has a value we need.
+  if (skipped_leaf_values > 0) { db->skip_values(skipped_leaf_values); }
+
+  while (!s->error && (s->input_value_count < s->num_input_values || s->src_pos < s->nz_count)) {
+    uint32_t target_pos;
+    uint32_t const src_pos = s->src_pos;
+
+    if (t < 64) {  // warp0..1
+      target_pos = min(src_pos + 2 * batch_size, s->nz_count + batch_size);
+    } else {       // warp2...
+      target_pos = min(s->nz_count, src_pos + batch_size);
+    }
+    __syncthreads();
+
+    // warp0 will decode the rep/def levels, warp1 will unpack a mini-batch of deltas.
+    // warp2 waits one cycle for warps 0/1 to produce a batch, and then stuffs values
+    // into the proper location in the output.
+    if (t < 32) {
+      // warp 0
+      // decode repetition and definition levels.
+      // - update validity vectors
+      // - updates offsets (for nested columns)
+      // - produces non-NULL value indices in s->nz_idx for subsequent decoding
+      gpuDecodeLevels<delta_rolling_buf_size, level_t>(s, sb, target_pos, rep, def, t);
+    } else if (t < 64) {
+      // warp 1
+      db->decode_batch();
+
+    } else if (t < 96 && src_pos < target_pos) {
+      // warp 2
+      // nesting level that is storing actual leaf values
+      int const leaf_level_index = s->col.max_nesting_depth - 1;
+
+      // process the mini-block in batches of 32
+      for (uint32_t sp = src_pos + lane_id; sp < src_pos + batch_size; sp += 32) {
+        // the position in the output column/buffer
+        int32_t dst_pos = sb->nz_idx[rolling_index<delta_rolling_buf_size>(sp)];
+
+        // handle skip_rows here. flat hierarchies can just skip up to first_row.
+        if (!has_repetition) { dst_pos -= s->first_row; }
+
+        // place value for this thread
+        if (dst_pos >= 0 && sp < target_pos) {
+          void* const dst = nesting_info_base[leaf_level_index].data_out + dst_pos * s->dtype_len;
+          if (s->dtype_len == 8) {
+            *static_cast<int64_t*>(dst) =
+              db->value[rolling_index<delta_rolling_buf_size>(sp + skipped_leaf_values)];
+          } else if (s->dtype_len == 4) {
+            *static_cast<int32_t*>(dst) =
+              db->value[rolling_index<delta_rolling_buf_size>(sp + skipped_leaf_values)];
+          }
+        }
+      }
+
+      if (lane_id == 0) { s->src_pos = src_pos + batch_size; }
+    }
+    __syncthreads();
+  }
+}
+
+}  // anonymous namespace
+
+/**
+ * @copydoc cudf::io::parquet::gpu::DecodeDeltaBinary
+ */
+void __host__ DecodeDeltaBinary(cudf::detail::hostdevice_vector<PageInfo>& pages,
+                                cudf::detail::hostdevice_vector<ColumnChunkDesc> const& chunks,
+                                size_t num_rows,
+                                size_t min_row,
+                                int level_type_size,
+                                rmm::cuda_stream_view stream)
+{
+  CUDF_EXPECTS(pages.size() > 0, "There is no page to decode");
```

Comment:
copied that from `DecodePageData`...would have to change it there as well.

Reply:
Let's re-imagine this as a TODO :)
What would be your preference?
I'm bringing this up because if we introduce a bug where this is called w/o pages in some corner case, a user would just see `There is no page to decode`, which is not very instructive. Instead, the reader could just work (correctly) :man_shrugging:

Reply:
Yeah, I've always been kind of puzzled by that test anyway.  I could at least change the error message to say no delta pages. Let me think on this.

Reply:
So there's an identical check in `reader::impl::decode_page_data()` (also in `reader::impl::read_chunk_internal()`), so the checks in the decode methods are both safe and unnecessary. I vote leave them alone for now.